ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/AttestationUtil.java

```diff
@@ -137,10 +163,104 @@ public static Bytes32 getAttestationMessageToSign(AttestationData attestationDat
-    return BeaconStateUtil.get_domain(
+    return get_domain(
             state.getFork(),
             BeaconStateUtil.slot_to_epoch(attestationData.getSlot()),
             Constants.DOMAIN_ATTESTATION)
         .intValue();
   }
+
+  public static boolean is_slashable_attestation_data(AttestationData data_1, AttestationData data_2){
+    //Check if ``data_1`` and ``data_2`` are slashable according to Casper FFG rules.
+    return (
+      //case 1: double vote || case 2: surround vote
+      (!data_1.equals(data_2) && data_1.getTarget_epoch().equals(data_2.getTarget_epoch()) ||
+      (data_1.getSource_epoch().compareTo(data_2.getSource_epoch()) < 0 && data_2.getTarget_epoch().compareTo(data_1.getTarget_epoch()) < 0))
+    );
+  }
+
+  public static IndexedAttestation convert_to_indexed(BeaconState state, Attestation attestation){
+    //Convert ``attestation`` to (almost) indexed-verifiable form.
+    List<UnsignedLong> attestations = get_attesting_indices(state, attestation.getData(), attestation.getAggregation_bitfield());
+    List<UnsignedLong>  custody_bit_1_indices = get_attesting_indices(state, attestation.getData(), attestation.getCustody_bitfield());
+
+    List<UnsignedLong>  custody_bit_0_indices = new ArrayList<UnsignedLong>();
+    for(int i = 0; i < attestations.size(); i++){
+      if(!custody_bit_1_indices.contains(UnsignedLong.valueOf(i)))custody_bit_0_indices.add(UnsignedLong.valueOf(i));
+    }
+    return new IndexedAttestation(
+            custody_bit_0_indices,
+            custody_bit_1_indices,
+            attestation.getData(),
+            attestation.getAggregate_signature()
+            );
+  }
+
+  public static List<UnsignedLong> get_attesting_indices(
+          BeaconState state, AttestationData attestation_data, Bytes bitfield) throws IllegalArgumentException {
+    //Return the sorted attesting indices corresponding to ``attestation_data`` and ``bitfield``.
+    List<Integer> committee = get_crosslink_committee(state, attestation_data.getTarget_epoch(), attestation_data.getCrosslink().getShard());
+    checkArgument(verify_bitfield(bitfield, committee.size()));
+    Iterator<Integer> itr = committee.iterator();
+
+    List<UnsignedLong> returnValue = new ArrayList<UnsignedLong>();
+    while(itr.hasNext()){
+      int index = itr.next().intValue();
+      int bitfield1 = get_bitfield_bit(bitfield, index);
+      if((get_bitfield_bit(bitfield, index) & 1) == 1) returnValue.add(UnsignedLong.valueOf(index));
+    }
+    return returnValue;
+  }
+
+  public static void validate_indexed_attestation(BeaconState state, IndexedAttestation indexed_attestation){
+    //Verify validity of ``indexed_attestation``.
+    List<UnsignedLong> bit_0_indices = indexed_attestation.getCustody_bit_0_indices();
+    List<UnsignedLong> bit_1_indices = indexed_attestation.getCustody_bit_1_indices();
+
+    //Verify no index has custody bit equal to 1 [to be removed in phase 1]
```

Comment:
I know this might be a lot of work, but can you move these comments above these checkArguments to the 2nd argument, so that when they throw we can specifically tell which one threw?